File: crawlerAndParser/Crawler.java
import java.io.File; import java.io.FileReader;
import java.io.BufferedReader;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.Queue;
import java.util.List;
import java.util.LinkedList;
import java.util.Set;
import java.util.HashSet;
import java.io.FileWriter;
import java.io.BufferedWriter;
import java.util.*;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.io.Writer;
import java.io.OutputStreamWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.File;
import java.net.HttpURLConnection;
import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.FileOutputStream;
import java.net.URL;
import java.net.URLEncoder;
import java.util.Map;
import java.net.URLDecoder;

/**
 * The class is written to download web pages from baike science related categories 
 * using POST request to get response containing url information for downloading
 * June 2017
 * e.g. http://baike.baidu.com/wikitag/taglist?tagId=75954
 * Change ROOT_DIR, and tagId etc to download pages from different categories
 * usage:
 * MAX_PAGE_NUM: use 1 for test, use 500 for crawling all pages
 * java Crawler tagId ROOT_DIR MAX_PAGE_NUM
 */

public class Crawler{

    //private static final int MAX_PAGE_NUM = 500;  // max respons page number, there are < 10000 items in any categroy
    private static final int BFS_DEPTH = 2;
    private static final int SLEEP_TIME_MS = 0; // 3000
    private static final String LIMIT_PARAM = "24";
    private static final String TIME_OUT = "3000";
    private static final String CONTENT_LENGTH = "40";


    // ROOT_DIR and tagId are passed from terminal input
    // change ROOT_DIR to be the folder where the output webpages to be stored
    //private static final String ROOT_DIR = "/home/jyu/data/baikeMedical/webpages/testdisease/";
    // change tagId to be the tagId which the source is
    //private static final String tagId = "75953"; // disease "68038"; // food

    private static String urlPost = "http://%s:8080/requests";
    private static final String REQUEST_URL = "http://baike.baidu.com/wikitag/api/getlemmas";
    private static final int BUFFER_SIZE = 4096;

    private static final String DOWNLOAD_ERROR_FILE = "/home/jyu/data/baikeSix/downloadFailed.txt";
    private static final String DOWNLOAD_STATS_FILE = "/home/jyu/data/baikeSix/downloadStats.txt";
    private static final String VISITED_FILE = "/home/jyu/data/baikeSix/visited.txt";
    private static final String URLS_BFS_FILE = "/home/jyu/data/baikeSix/urlsBfs.txt";
                       
    public Crawler(){}
    public Crawler(String url){
        setUrlPost(url);
    }
 
    /**
     * Downloads a file from a URL
     * @param fileURL HTTP URL of the file to be downloaded
     * @param ROOT_DIR path of the directory to save the file
     * @throws IOException
     */
    private static int downloadFile(String fileURL, String ROOT_DIR)
            throws IOException {
	//System.out.println("downloadFile fileURL: " + fileURL);
	//System.out.println("downloadFile ROOT_DIR: " + ROOT_DIR);
        URL url = new URL(fileURL);
        HttpURLConnection httpConn = (HttpURLConnection) url.openConnection();
        int responseCode = httpConn.getResponseCode();
 
	// 
	boolean redirect = false;
	if(responseCode != HttpURLConnection.HTTP_OK){
		if(responseCode == HttpURLConnection.HTTP_MOVED_TEMP
		    || responseCode == HttpURLConnection.HTTP_MOVED_PERM
		        || responseCode == HttpURLConnection.HTTP_SEE_OTHER)
	        redirect = true;
	}
	if(redirect){
		// get redicret url from "location" header field
		String newUrl = httpConn.getHeaderField("Location");
		// open the new connection again
		System.out.println(newUrl);
		httpConn = (HttpURLConnection) new URL(newUrl).openConnection();

		System.out.println("Redirect to URL: " + newUrl);
	}
	responseCode = httpConn.getResponseCode();

	if(responseCode == 302){
		String newUrl = httpConn.getHeaderField("Location");
		httpConn = (HttpURLConnection) new URL(newUrl).openConnection();
		responseCode = httpConn.getResponseCode();
	}

        // always check HTTP response code first
        if (responseCode == HttpURLConnection.HTTP_OK) {
            String fileName = fileURL.substring(fileURL.lastIndexOf("/") + 1,
                        fileURL.length());

            // opens input stream from the HTTP connection
            InputStream inputStream = httpConn.getInputStream();
            String saveFilePath = ROOT_DIR + File.separator + fileName;
             
            // opens an output stream to save into file
            FileOutputStream outputStream = new FileOutputStream(saveFilePath);
 
            int bytesRead = -1;
            byte[] buffer = new byte[BUFFER_SIZE];
            while ((bytesRead = inputStream.read(buffer)) != -1) {
                outputStream.write(buffer, 0, bytesRead);
            }
            outputStream.close();
            inputStream.close();
            System.out.println("File downloaded");

	}else{
		String errorMessage = "No file to download. Server replied HTTP code: " + responseCode;
		appendToFile(fileURL + ' ' + errorMessage + '\n', DOWNLOAD_ERROR_FILE);
		System.out.println(errorMessage);
		return 0;
        }
        httpConn.disconnect();
        return 1;
    }


    private void setUrlPost(String urlPost) {
        this.urlPost = urlPost;
    }

    private void setMsgToSend(String msg) {
        this.msgToSend = msg;
    }
    String msgToSend = "";

    private String sendPost(String msg) {
        String res = "";
        msgToSend = msg;
        try {
            res = sendPost();
        } catch (Exception e) {
            //LOGGER.error("error post data " + msg + " to " + urlPost, e);
        }
        return res;
    }
    // HTTP POST request
    private String sendPost() throws Exception {
        if (urlPost == null) {
            //LOGGER.info("skip post " + msgToSend);
        }
        //LOGGER.info("sending " + msgToSend + " to " + urlPost);

        URL obj = new URL(urlPost);
        HttpURLConnection con = (HttpURLConnection) obj.openConnection();

        //add reuqest header
        con.setRequestMethod("POST");
        con.setRequestProperty("User-Agent", "");
        //con.setRequestProperty("Accept-Language", "en-US,en;q=0.5");

        // Send post request
        con.setDoOutput(true);
        con.setRequestProperty("encoding", "utf-8");
        DataOutputStream wr = new DataOutputStream(con.getOutputStream());

        //wr.writeBytes(msgToSend);  // luanma
        wr.write(msgToSend.getBytes());
        wr.flush();
        wr.close();

        int responseCode = con.getResponseCode();
        //LOGGER.debug("Sending 'POST' request to URL : " + urlPost);
        String log = msgToSend;
        BufferedReader in = new BufferedReader(
                new InputStreamReader(con.getInputStream()));
        String inputLine;
        StringBuffer response = new StringBuffer();

        while ((inputLine = in.readLine()) != null) {
            response.append(inputLine);
        }
        in.close();

        //print result
        String res = response.toString();
        return res;
    }

    public static String sendGet(String url) throws Exception {
        // HTTP GET request\     String urlGet = "http://%s:8080/requests?bizuin=";
        URL obj = new URL(url);
        HttpURLConnection con = (HttpURLConnection) obj.openConnection();

        // optional default is GET
        con.setRequestMethod("GET");

        //add request header
        con.setRequestProperty("User-Agent", "");

        int responseCode = con.getResponseCode();
        //LOGGER.info("\nSending 'GET' request to URL : " + url);
        //LOGGER.info("Response Code : " + responseCode);

        BufferedReader in = new BufferedReader(
                new InputStreamReader(con.getInputStream()));
        String inputLine;
        StringBuffer response = new StringBuffer();

        while ((inputLine = in.readLine()) != null) {
            response.append(inputLine);
        }
        in.close();

        //LOGGER.info(response.toString());
        return response.toString();
    }

    private static void writeToFile(List<String> strs, String filePath){
        try(PrintWriter printWriter = new PrintWriter(filePath)){
            for(String s : strs){
                printWriter.println(s);
            }
        }catch(Exception e){
            System.out.printf("Can't write: %s!/n", filePath);
        }
    }

   private static List<String> getUrls(List<String> responses){
        // process urls and write id name map file
        List<String> urls = new LinkedList<String>();
        for(String res : responses){
            for(int i = 0; i < res.length()-4; i++){
              if(res.substring(i, i+4).equals("http")){
                   int j = i+4;
                   while(res.charAt(j) != '"'){
                       j++;
                   }
                   String urlRaw = res.substring(i, j);
                   StringBuilder urlBuilder = new StringBuilder();
                   for(char c : urlRaw.toCharArray()){
                       if(c != '\\'){
                           urlBuilder.append(c);
                        }
                   }
		   String url = urlBuilder.toString();
                   if(url != null && !url.isEmpty() && url.length() > 12 &&
				   url.substring(0, 12).equals("http://baike")){
			   urls.add(url);
		   }
              }
            }
        }
        return urls;
   }

   private static List<String> getNameIds(List<String> urls){
        List<String> nameIds = new LinkedList<String>();
       for(String url : urls){
           String[] urlSplit = url.split("/");
           if(urlSplit.length < 6){
               //System.out.println("getNameIds: " + url);
           }else{
               String id = urlSplit[5]; 
               try{
                   String name = URLDecoder.decode(urlSplit[4], "UTF-8"); 
                   //System.out.println(name+" "+id); 
                   nameIds.add(name + " " + id);
               }catch(Exception e){
                   //System.out.println("Can't decode url with utf8!");
               }
           }
       }
       return nameIds;
   }

   private static List<String> getPostRequestResponses(Crawler crawler, String tagId, int MAX_PAGE_NUM){
        // get response from post request
        List<String> responses = new LinkedList<String>();
        for(int pageNum = 0; pageNum < MAX_PAGE_NUM; pageNum++){
            String params = "limit=" + LIMIT_PARAM + "&timeout=" + TIME_OUT + "&filterTags=%5B%5D&tagId=" + 
		    tagId +"&fromLemma=false&contentLength="+ CONTENT_LENGTH + "&page="+pageNum;
            String res = crawler.sendPost(params);
            responses.add(res);
	    System.out.println("Getting post response page: " + pageNum);
        }
        return responses;
   }


    public static void appendToFile(String data, String filePath) {
		BufferedWriter bw = null;
		FileWriter fw = null;
		try {
			File file = new File(filePath);
			// if file doesnt exists, then create it
			if (!file.exists()) {
				file.createNewFile();
			}
			// true = append file
			fw = new FileWriter(file.getAbsoluteFile(), true);
			bw = new BufferedWriter(fw);
			bw.write(data);
			System.out.println("Done");
		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			try {
				if (bw != null)
					bw.close();
				if (fw != null)
					fw.close();
			} catch (IOException ex) {
				ex.printStackTrace();
			}
		}
	}


    // get urls recursively by BFS with depth of 3 level
    private static Set<String> getUrlsBfs(String url_start, int depth){

	Set<String> urls = new HashSet<>();
        // timeout connection after 500 miliseconds
        System.setProperty("sun.net.client.defaultConnectTimeout", "500");
        System.setProperty("sun.net.client.defaultReadTimeout",    "1000");

        // list of web pages to be examined
        Queue<String> queue = new LinkedList<String>();
        queue.add(url_start);

        // set of examined web pages
        Set<String> marked = new HashSet<String>();
        marked.add(url_start);

        // breadth first search crawl of web
        int level = 0;
        while (level < depth && !queue.isEmpty()) {
            level++;
            String v = queue.remove();
            System.out.println(v);

            String input = null;
            try {
                In in = new In(v);
                input = in.readAll();
            }
            catch (IllegalArgumentException e) {
                System.out.println("[could not open " + v + "]");
                continue;
            }
            // if (input == null) continue;

            String regexp = "/item/[-%\\w]+(\\/\\d+)?";
            Pattern pattern = Pattern.compile(regexp);

            Matcher matcher = pattern.matcher(input);

            // find and print all matches
            while (matcher.find()) {
                String w = matcher.group();
		String newUrl = "http://baike.baidu.com" + w;
                if (!marked.contains(w)) {
                    queue.add(w);
                    marked.add(w);
		    urls.add(newUrl);
                }
            }

        }
	return urls;
    }

    private static Set<String> getVisited(String visitedFile){
	    Set<String> visited = new HashSet<>();
	    try{
		    FileReader fr = new FileReader(VISITED_FILE);
		    BufferedReader br = new BufferedReader(fr);
		    try{
			    String line = br.readLine();
			    while(line != null){
				    visited.add(line.trim());
				    line = br.readLine();
			    }
		    }finally{
			    br.close();
		    }
	    }catch(IOException e){
		    System.out.println("visited file not exists!");
	    }
	    return visited;
    }

    private static void saveVisited(Set<String> urls){
	    try{
		    FileWriter fw = new FileWriter(VISITED_FILE, true);
		    BufferedWriter out = new BufferedWriter(fw);
		    Iterator it = urls.iterator();
		    while(it.hasNext()){
			    out.write(it.next()+"\n");
		    }
		    out.close();
	    }catch(IOException e){
		    System.out.println("can't write to visited file");
            }
    }

    private static void saveUrlsBfs(List<String> urlsBfsOnelevel){
	    try{
		    FileWriter fw = new FileWriter(URLS_BFS_FILE, true);
		    BufferedWriter out = new BufferedWriter(fw);
		    Iterator it = urlsBfsOnelevel.iterator();
		    while(it.hasNext()){
			    out.write(it.next()+"\n");
		    }
		    out.close();
	    }catch(IOException e){
		    System.out.println("can't write to urlsBfsOnelevel file");
            }
    }


    public static void main(String[] args) throws Exception {
        Crawler crawler = new Crawler();
        crawler.setUrlPost(REQUEST_URL);

	String tagId = args[0];
	String ROOT_DIR = args[1];
        String title = args[2];
	int MAX_PAGE_NUM = Integer.parseInt(args[2]);
	System.out.println("tagId: " + tagId);
	System.out.println("ROOT_DIR: " + ROOT_DIR);
	System.out.println("MAX_PAGE_NUM: " + MAX_PAGE_NUM);

        File directory = new File(ROOT_DIR);
        if(!directory.exists()){
            directory.mkdir();
        }

        File directory_saveDir = new File(ROOT_DIR + title);
        if(!directory_saveDir.exists()){
            directory_saveDir.mkdir();
        }

        List<String> responses = getPostRequestResponses(crawler, tagId, MAX_PAGE_NUM);
        List<String> urls = getUrls(responses);
        List<String> nameIds = getNameIds(urls);
	List<String> urlsBfsOnelevel = new LinkedList<String>();
        
        writeToFile(responses, ROOT_DIR+"/responses.txt");
        writeToFile(urls, ROOT_DIR+"/urls.txt");
        writeToFile(nameIds, ROOT_DIR+"/nameIdMap.txt");

	Set<String> visited = getVisited(VISITED_FILE);

        // download webpages
	int count_attempt = 0;
	int count_downloaded = 0;
        for(String url_start : urls){
		Set<String> urlsBfs = getUrlsBfs(url_start, BFS_DEPTH);
		for(String url : urlsBfs){
			String[] urlSplit = url.split("/");
			if(urlSplit.length < 4){
			       System.out.println("Wired: url split len < 4: " + url_start);	
			}else{
				String doc_title = url.split("/")[4];
			        if(!visited.contains(doc_title)){
			        	urlsBfsOnelevel.add(url);
				}
			}
			String[] url_startSplit = url_start.split("/");
			if(url_startSplit.length < 4){
			       System.out.println("Wired: url split len < 4: " + url_start);	
			}else{
				visited.add(url_startSplit[4]);
			}
			count_attempt = count_attempt + 1;
			System.out.println("Downloading url: " + url_start);
			System.out.println("Saving page to ROOT_DIR: " + ROOT_DIR + title);
			count_downloaded += downloadFile(url_start, ROOT_DIR + title);
			Thread.sleep(SLEEP_TIME_MS);
		}
	//	}
           // }else{
                //System.out.println("main: " + url);
            //}
        }
        String counts = "tagId: " + tagId + " " 
	        + "ROOT_DIR: " + ROOT_DIR + " "	
                + "title: " + title + " "
		+ "Attempts: " +  count_attempt + " " + "Downloaded:" + count_downloaded + "\n";
        appendToFile(counts, DOWNLOAD_STATS_FILE);
	saveVisited(visited);
	saveUrlsBfs(urlsBfsOnelevel);
    }
}

File: crawlerAndParser/crawler.py
# August 2017
# ==============================================================================
"""
crawls webpages with some medical label starting from baidu baike science using BFS

"""
import sys
import _thread
import os
import re
from bs4 import BeautifulSoup
from urllib import request, parse
from urllib.request import urlopen, Request
from urllib.parse import unquote
import pathlib
import json

DIR_OUT = '/home/jyu/data/baike/scienceAndMedicalBfs/'
LEMMA_DICT_FILE = '/home/jyu/data/baike/lemma_dict.txt'
URL_ROOT = 'https://baike.baidu.com/science'
URL_PRE = 'https://baike.baidu.com/item/'
lables_medical={'医学', '医疗', '医药', '药物', '诊疗', '疾病', '中药', '医学术语', '病毒'} 
DEPTH_BFS = 10000 # set it to be big to crawl all medical baike pages


# get science category tags' titles, ids
# ==============================================================================
def getTags(url_root):
    with urlopen(url_root) as conn:
        soup = BeautifulSoup(conn.read(), 'html5lib')
    tags = []
    for tag in soup.find_all('h4'):
        url = tag.a.attrs['href']
        url_splited = url.split('tagId=')
        titles = []
        titles.append(tag.get_text())
        if len(url_splited) == 2:
            tag_id = url_splited[-1]
            tags.append((tag_id, titles))
        else:
            tags.extend(getTags_medical(url, titles))
    return tags

def getTags_medical(url_root, titles_root):
    with urlopen(url_root) as conn:
        soup = BeautifulSoup(conn.read(), 'html5lib')
    tags = []
    tag_box = soup.find('ul', attrs={'class':'knowledge-list cmn-clearfix'}).find('li')
    while tag_box != None: 
        titles = list(titles_root)
        titles.append(tag_box.span.get_text())
        url = tag_box.a.attrs['href']
        url_splitted = url.split('tagId=')
        if len(url_splitted) == 2:
            tag_id = url_splitted[-1]
            tags.append((tag_id, titles))
        tag_box = tag_box.find_next_sibling()
    return tags


# get lemma list for science category tags
# ==============================================================================
def get_lemmaList(tagId, test_flag):
    url = 'https://baike.baidu.com/wikitag/api/getlemmas'
    if test_flag:
        totalPage = 0#1000 #1 for testing
        limit = '1'#'24' #1 for testing, 24
    else:
        totalPage = 1000
        limit = '24'
    page_id = 0
    lemmaList = []
    while page_id <= totalPage:
        #print("page_id: " + str(page_id) + ' of total ' + str(totalPage))
        data = {'limit':limit, 'timeout':'3000','filterTags':'%5B%5D','tagId':tagId,'fromLemma':'false','contentLength':'40','page':page_id}
        data = parse.urlencode(data).encode()
        req = Request(url, data = data)
        resp = urlopen(req)
        if resp.status != 200:
            print(tagId, page_id, resp.status, resp.reason)
            continue
        resp_json = json.loads(resp.read())
        lemmaList.extend(resp_json['lemmaList'])
        if test_flag:
            pass
        else:
            totalPage = resp_json['totalPage']  #comment this line for testing
        page_id = page_id + 1
    total = resp_json['total']
    #print('total: ' + str(total))
    #print('len of lammaList: ' + str(len(lemmaList)))
    return lemmaList, total


# write json data to file
# ==============================================================================
def write_to_file(data, file_path):
    directory = os.path.dirname(file_path)
    pathlib.Path(directory).mkdir(parents=True, exist_ok=True) 
    with open(file_path, 'w') as outfile:
        json.dump(data, outfile, ensure_ascii=False)


# download pages using BFS with depth
# ==============================================================================
def download_bfs(queue, depth, out_dir):
    p = re.compile('"/item/\S+[/\d+]*"')
    level = 0
    lables = set()
    visited = set(queue) 
    while queue: 
        level += 1
        queue_cur = list(queue)
        queue = []
        for item in queue_cur:
           # download page
           url = URL_PRE + item
           response = urlopen(url)
           if response.status != 200:
               print(response.status)
           web_content = response.read()
           file_path = os.path.join(out_dir, unquote(item))
           #print('file_path: ' + str(file_path))
           f = open(file_path, 'wb')
           f.write(web_content)
           f.close
           print('level: ' + str(level) + '\n' + 'depth: ' + str(depth))
           # add valid child note to queue
           if level <= depth:
                soup = BeautifulSoup(web_content, 'html5lib')
                main_content = soup.find('div', attrs={'class':'main-content'})
                item_list = p.findall(str(main_content))
                for item in item_list:
                    item = item.strip().strip('"').strip('/item/')
                    url_new = URL_PRE + item
                    #print('url_new: ' + url_new)
                    tag_boxes = BeautifulSoup(urlopen(Request(url_new)).read(), 'html5lib').find_all('span', attrs={'class':'taglist'})
                    for tag_box in tag_boxes:
                        tag_text = tag_box.get_text().strip()
                        lables.add(tag_text)
                        if tag_text in lables_medical or '医学' in tag_text or '疾病' in tag_text or '医药' in tag_text or '医疗' in tag_text or '诊疗' in tag_text or '药物' in tag_text or '健康' in tag_text:
                            if url_new not in visited:
                                queue.append(item)
                                visited.add(item)
                                print(tag_text + ' is in lables_medical and appending ' + url_new + ' to queue!')
    return visited, lables


if __name__ == '__main__':
    
    try:
        input_arg = sys.argv[1]    
        if input_arg == 'test':
            test_flag = True
    except:
        test_flag = False
        #print('Please add test flag, True or False!')

    if test_flag:
        DIR_OUT = os.path.join(DIR_OUT, 'test')
        DEPTH_BFS = 1
    os.makedirs(DIR_OUT, exist_ok=True) 
    
    # get tags list
    tag_id_titles_list = getTags(URL_ROOT)
    print('tag_id_titles_list: ' + str(tag_id_titles_list))   

    # get lemma diction
    lemma_dict = {}
    for (tag_id, title) in tag_id_titles_list:
        print("processing tag_id: " + tag_id)
        lemmaList, total = get_lemmaList(tag_id, test_flag) 
        lemma_dict[tag_id]={}
        lemma_dict[tag_id]['tag_title'] = title
        lemma_dict[tag_id]['lemmal_list'] = lemmaList 
        lemma_dict[tag_id]['total'] = total
    write_to_file(lemma_dict, LEMMA_DICT_FILE)
    
    # download pages from lemma diction
    with open(LEMMA_DICT_FILE) as f:
        lemma_dict = json.load(f)
    queue = []
    num_items = 0
    for tag_id in lemma_dict.keys():
        queue_sub = []
        lemma_list = lemma_dict[tag_id]['lemmal_list']
        num_items += lemma_dict[tag_id]['total']
        num_items_sub = lemma_dict[tag_id]['total']
        for lemma in lemma_list:
            url = lemma['lemmaUrl']
            #print('lemma: ' + str(lemma))
            #print('url: ' + url)
            try:
                url_split = url.split('/')
                if len(url_split) >= 5:
                    item = url_split[4]
                else:
                    print('main: can\'t get item name from url: ' + url)
            except:
                item = lemma['lemmaTitle']
                print('get item name from lemmaTitle instead!: ' + item)
            queue.append(item)
            #queue_sub.append(item)
        #print('num_items_sub: ' + str(num_items_sub))
        #print('len of queue_sub: ' + str(len(queue_sub)))
    print('num_items: ' + str(num_items))
    print('len of queue: ' + str(len(queue)))
    dir_out_webpages = os.path.join(DIR_OUT, 'webpages')
    os.makedirs(dir_out_webpages, exist_ok=True)
    visited, lables = download_bfs(queue, DEPTH_BFS, dir_out_webpages)

    VISITED_FILE = os.path.join(DIR_OUT, 'visited.txt')
    LABLES_FILE = os.path.join(DIR_OUT, 'lables.txt')
    write_to_file(list(lables), LABLES_FILE)
    write_to_file(list(visited), VISITED_FILE)
